README.md
# Serverless Image Handler Readme

The serverless-image-handler mono repository contains the source code and documentation for the [image-handler](#image-handler) and [image-thumbs](#image-thumbs) AWS Lambda functions.

## Table of Contents

- [Serverless Image Handler Readme](#serverless-image-handler-readme)
- [Table of Contents](#table-of-contents)
- [Architecture](#architecture)
- [Image Handler](#image-handler)
    - [Overview](#overview)
    - [Prerequisites](#prerequisites)
    - [Usage](#usage)
    - [Building](#building)
    - [Testing](#testing)
    - [Infrastructure deployment](#infrastructure-deployment)
- [Image Thumbs](#image-thumbs)
    - [Overview](#overview-1)
    - [Prerequisites](#prerequisites-1)
    - [Building](#building-1)
    - [Test / Invoke](#test--invoke)
    - [Infrastructure deployment](#infrastructure-deployment-1)

### Architecture

![Architecture](architecture.png)

## Image Handler

### Overview

The Image-handler is a solution Image Handler is a serverless image processing project built with Node.js 18 and the [Sharp](https://sharp.pixelplumbing.com/en/stable/) library. 
It allows you to dynamically scale images on the fly and serves them through AWS Lambda, AWS CloudFront, and AWS S3.

It aws originally forked from [aws-solutions/serverless-image-handler repository](https://github.com/aws-solutions/serverless-image-handler), but has been heavily modified to suit our needs.

### Prerequisites

- [Node.js](https://nodejs.org/en/) v20.x or later
- [Terraform](https://www.terraform.io/downloads.html)
- Make / npm
- AWS credentials with sufficient permissions to read images from S3

### Usage

To scale images dynamically on the fly, you can make HTTP requests to the AWS CloudFront distribution URL. 
The images are fetched from AWS S3, processed using Node.js 20 and Sharp / libvips, and then served through CloudFront.

Example URL:

```https://images.t-online.de/4k_hdr.jpg```

For more details see the [Usage](docs/Usage.md) documentation.

### Environment variables

The following environment variables are used by the image-handler:

| Name                      | Description                                     |
|---------------------------|-------------------------------------------------|
| `AUTO_WEBP`               | Flag if the AUTO WEBP feature should be enabled |
| `CORS_ENABLED`            | Flag if CORS should be enabled                  |
| `CORS_ORIGIN`             | CORS origin.                                    |
| `SOURCE_BUCKETS`          | S3 Bucket with source images                    |

### Building

To build the package run:

```make FUNC=image-handler build```

### Testing

Run tests using the following Make command:

```make npm/test```

### Infrastructure deployment

Deploy the infrastructure using Terraform with the following Make command:

```make FUNC=image-handler tf```

## Special interest section

Useful links:

* Image too large with default settings: `2023/02/JPCPt616git7/image.png`

## Ströer specific adaptations

### The file name does not matter

The way we process our images does not depend on the file name. The part that copies the images from 
the CMS into our delivery bucket will rename all images to `image.${extension}`.

When rendering the image, the image-handler will always look for this filename, so it does not matter
what the original file name was. e.g. 

- `2023/02/JPCPt616git7/seo-title.png` will be served as `image.png`
- `2023/02/JPCPt616git7/other-usage-different-title.png` will be served as `image.png`

resulting in the same response.

### Cropping coordinates format

- the original solution introduced thumbor cropping in [v6](https://github.com/aws-solutions/serverless-image-handler/commit/76558b787b9417450ee4a4f19dc9548be6dbada7)
- we have implemented this feature in advance, but with a slightly different syntax:
  - The original solution uses `crop=left,top,right,bottom`
  - We use `crop=left,top,width,height` to be more consistent with the CMS

### Expired Content

- The original solution uses the `Expires` header to cache images in the browser
- Our solution:
  - reduces the `Cache-Control: max-age` according to the `Expires` header to handle expired content
  - Also, once the Expires header is reached, the image-handler will return a `http/410 (GONE)` status code 

### Additional filter: thumbhash

- `/filters:thumbhash()/` will trigger a hash-based thumbnail generation. See https://evanw.github.io/thumbhash/
- The image can be cropped prior to this filter, but will ultimately be resized to `100x100` pixel to generate an efficient thumbnail
- The response will be `base64` encoded binary that can be converted via `Thumbhash.thumbHashToDataURL()`

### Next data workaround

Within the HTML markup resides the `next-data` [attribute](https://github.com/vercel/next.js/discussions/15117)

Within this data structure there are some image URLs that are not directly intended to be rendered. Consider 
them as templates. 
Google (and other bots) will find and use this URL nonetheless. This image-handler will replace all calls that
contain template variables `/__WIDTH__x0/` with an actual size to mitigate this.

### Removal of features not required

If required once more, they need to be pulled from the original repository:

- Recognition and the corresponding code:
  - "Smart Crop" (aka face recognition)
  - Content Moderation (e.g. NSFW detection)
- Watermarking/Overlaying
- Secretsmanager and URL signing
- Dynamic S3 bucket selection

### Updating Sharp version

You need to keep the sharp version in `package.json` and the lambda layer in sync.

First, create a new lambda layer version with the updated sharp version.

```terraform
locals {
  sharp_version = "0.34.4"
}
```

Then update the function code and create a new lambda version.

CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [5.1.0] - 2020-11-19

### ⚠ BREAKING CHANGES

- **Image URL Signature**: When image URL signature is enabled, all URLs including existing URLs should have `signature` query parameter.

### Added

- Image URL signature: [#111](https://github.com/awslabs/serverless-image-handler/issues/111), [#203](https://github.com/awslabs/serverless-image-handler/issues/203), [#221](https://github.com/awslabs/serverless-image-handler/issues/221), [#227](https://github.com/awslabs/serverless-image-handler/pull/227)
- AWS Lambda `413` error handling. When the response payload is bigger than 6MB, it throws `TooLargeImageException`: [#35](https://github.com/awslabs/serverless-image-handler/issues/35), [#97](https://github.com/awslabs/serverless-image-handler/issues/97), [#193](https://github.com/awslabs/serverless-image-handler/issues/193), [#204](https://github.com/awslabs/serverless-image-handler/issues/204)
- Default fallback image: [#137](https://github.com/awslabs/serverless-image-handler/issues/137)
- Unit tests for custom resource: `100%` coverage
- Add `SVG` support. When any edits are used, the output would be automatically `PNG` unless the output format is specified: [#31](https://github.com/awslabs/serverless-image-handler/issues/31), [#234](https://github.com/awslabs/serverless-image-handler/issues/234)
- Custom headers: [#182](https://github.com/awslabs/serverless-image-handler/pull/182)
- Enabling ALB Support : [#201](https://github.com/awslabs/serverless-image-handler/pull/201)

### Fixed

- Thumbor paths broken if they include "-" and "100x100": [#208](https://github.com/awslabs/serverless-image-handler/issues/208)
- Rewrite doesn't seem to be working: [#121](https://github.com/awslabs/serverless-image-handler/issues/121)
- Correct EXIF: [#197](https://github.com/awslabs/serverless-image-handler/issues/197), [#220](https://github.com/awslabs/serverless-image-handler/issues/220), [#235](https://github.com/awslabs/serverless-image-handler/issues/235), [#236](https://github.com/awslabs/serverless-image-handler/issues/236), [#240](https://github.com/awslabs/serverless-image-handler/issues/240)
- Sub folder support in Thumbor `watermark` filter: [#231](https://github.com/awslabs/serverless-image-handler/issues/231)

### Changed

- AWS CDK and AWS Solutions Constructs version (from 1.57.0 to 1.64.1)
- sharp base version (from 0.25.4 to 0.26.1)
- Refactors the custom resource Lambda source code
- Migrate unit tests to use `jest`
- Move all `aws-sdk` in `ImageHandler` Labmda function to `index.js` for the best practice
- Enhance the default error message not to show empty JSON: [#206](https://github.com/awslabs/serverless-image-handler/issues/206)

### Removed

- Remove `manifest-generator`

## [5.0.0] - 2020-08-31

### Added

- AWS CDK and AWS Solutions Constructs to create AWS CloudFormation template

### Fixed

- Auto WebP does not work properly: [#195](https://github.com/awslabs/serverless-image-handler/pull/195), [#200](https://github.com/awslabs/serverless-image-handler/issues/200), [#205](https://github.com/awslabs/serverless-image-handler/issues/205)
- A bug where base64 encoding containing slash: [#194](https://github.com/awslabs/serverless-image-handler/pull/194)
- Thumbor issues:
  - `0` size support: [#183](https://github.com/awslabs/serverless-image-handler/issues/183)
  - `convolution` filter does not work: [#187](https://github.com/awslabs/serverless-image-handler/issues/187)
  - `fill` filter does not work: [#190](https://github.com/awslabs/serverless-image-handler/issues/190)
- **Note that** duplicated features has been merged gracefully.

### Removed

- AWS CloudFormation template: `serverless-image-handler.template`

### Changed

- sharp base version (from 0.23.4 to 0.25.4)
- Remove `Promise` to return since `async` functions return promises: [#189](https://github.com/awslabs/serverless-image-handler/issues/189)
- Unit test statement coverage improvement:
  - `image-handler.js`: `79.05%` to `100%`
  - `image-request.js`: `93.58%` to `100%`
  - `thumbor-mapping.js`: `99.29%` to `100%`
  - `overall`: `91.55%` to `100%`

## [4.2] - 2020-02-06

### Added

- Honor outputFormat Parameter from the pull request [#117](https://github.com/awslabs/serverless-image-handler/pull/117)
- Support serving images under s3 subdirectories, Fix to make /fit-in/ work; Fix for VipsJpeg: Invalid SOS error plus several other critical fixes from the pull request [#130](https://github.com/awslabs/serverless-image-handler/pull/130)
- Allow regex in SOURCE_BUCKETS for environment variable from the pull request [#138](https://github.com/awslabs/serverless-image-handler/pull/138)
- Fix build script on other platforms from the pull request [#139](https://github.com/awslabs/serverless-image-handler/pull/139)
- Add Cache-Control response header from the pull request [#151](https://github.com/awslabs/serverless-image-handler/pull/151)
- Add AUTO_WEBP option to automatically serve WebP if the client supports it from the pull request [#152](https://github.com/awslabs/serverless-image-handler/pull/152)
- Add AUTO_AVIF option to automatically serve AVIF if the client supports it
- Use HTTP 404 & forward Cache-Control, Content-Type, Expires, and Last-Modified headers from S3 from the pull request [#158](https://github.com/awslabs/serverless-image-handler/pull/158)
- fix: DeprecationWarning: Buffer() is deprecated from the pull request [#174](https://github.com/awslabs/serverless-image-handler/pull/174)
- Add hex color support for Thumbor `filters:background_color` and `filters:fill` [#154](https://github.com/awslabs/serverless-image-handler/issues/154)

source/image-handler/image-request.js
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

const logger = require("./logger");
const ThumborMapping = require("./thumbor-mapping");

class ImageRequest {
  constructor(s3) {
    this.s3 = s3;
  }

  /**
   * Initializer function for creating a new image request, used by the image
   * handler to perform image modifications.
   * @param {object} event - Lambda request body.
   */
  async setup(event) {

    this.requestType = this.parseRequestType(event);
    this.bucket = this.parseImageBucket(event, this.requestType);
    this.key = this.parseImageKey(event, this.requestType);
    this.edits = this.parseImageEdits(event, this.requestType);
    this.cropping = this.parseCropping(event, this.requestType);
    this.originalImage = await this.getOriginalImage(this.bucket, this.key);
    this.headers = this.parseImageHeaders(event, this.requestType);
    this.isAlb = event.requestContext && event.requestContext.hasOwnProperty("elb");

    if (!this.headers) {
      delete this.headers;
    }

    // If the original image is SVG file and it has any edits but no output format, change the format to WebP.
    if (this.ContentType === "image/svg+xml" &&
      this.edits &&
      Object.keys(this.edits).length > 0 &&
      !this.edits.toFormat) {
      this.outputFormat = "png";
    }

    /* Decide the output format of the image.
     * 1) If the format is provided, the output format is the provided format.
     * 2) If headers contain "Accept: image/webp", the output format is webp.
     * 3) Use the default image format for the rest of cases.
     */
    if (this.ContentType !== 'image/svg+xml' || this.edits.toFormat || this.outputFormat) {
      let outputFormat = this.getOutputFormat(event);
      if (this.edits && this.edits.toFormat) {
        this.outputFormat = this.edits.toFormat;
      } else if (outputFormat) {
        this.outputFormat = outputFormat;
      }
    }

    // Fix quality for Thumbor and Custom request type if outputFormat is different from quality type.
    if (this.outputFormat) {
      const requestType = ["Custom", "Thumbor"];
      const acceptedValues = ["jpeg", "png", "webp", "tiff", "heif"];

      this.ContentType = `image/${this.outputFormat}`;
      if (
        requestType.includes(this.requestType) &&
        acceptedValues.includes(this.outputFormat)
      ) {
        let qualityKey = Object.keys(this.edits).filter((key) =>
          acceptedValues.includes(key)
        )[0];
        if (qualityKey && qualityKey !== this.outputFormat) {
          const qualityValue = this.edits[qualityKey];
          this.edits[this.outputFormat] = qualityValue;
          delete this.edits[qualityKey];
        }
      }
    }

    delete this.s3;

    return this;
  }

  /**
   * Gets the original image from an Amazon S3 bucket.
   * @param {string} bucket - The name of the bucket containing the image.
   * @param {string} key - The key name corresponding to the image.
   * @return {Promise} - The original image or an error.
   */
  async getOriginalImage(bucket, key) {
    const imageLocation = {Bucket: bucket, Key: key};
    try {
      const originalImage = await this.s3.getObject(imageLocation).promise();
      const metaData = originalImage['Metadata'];
      const isGone = metaData && metaData['buzz-status-code'] && metaData['buzz-status-code'] === '410'

      if (originalImage.ContentType) {
        // If using default s3 ContentType infer from hex headers
        if (originalImage.ContentType === 'binary/octet-stream') {
          const imageBuffer = Buffer.from(originalImage.Body);
          this.ContentType = this.inferImageType(imageBuffer);
        } else {
          this.ContentType = originalImage.ContentType;
        }
      } else {
        this.ContentType = "image";
      }

      if (originalImage.Expires) {
        this.Expires = new Date(originalImage.Expires);
      } else if (isGone) {
        logger.warn(`Content ${imageLocation} is gone`)
        this.Expires = new Date(0);
      }

      if (originalImage.LastModified) {
        this.LastModified = new Date(originalImage.LastModified);
      }

      if (originalImage.CacheControl) {
        this.CacheControl = originalImage.CacheControl;
      } else {
        this.CacheControl = "max-age=31536000,public";
      }

      if (originalImage.ETag) {
        this.ETag = originalImage.ETag;
      }

      return originalImage.Body;
    } catch (err) {
      throw {
        status: "NoSuchKey" === err.code ? 404 : 500,
        code: (err.code).toString(),
        message: err.message
      };
    }
  }

  /**
   * Parses the name of the appropriate Amazon S3 bucket to source the
   * original image from.
   * @param {string} event - Lambda request body.
   * @param {string} requestType - Image handler request type.
   */
  parseImageBucket(event, requestType) {
    if (requestType === "Default") {
      // Decode the image request
      const decoded = this.decodeRequest(event);
      if (decoded.bucket !== undefined) {
        // Check the provided bucket against the allowed list
        const sourceBuckets = this.getAllowedSourceBuckets();
        if (
          sourceBuckets.includes(decoded.bucket) ||
          decoded.bucket.match(new RegExp("^" + sourceBuckets[0] + "$"))
        ) {
          return decoded.bucket;
        } else {
          throw {
            status: 403,
            code: "ImageBucket::CannotAccessBucket",
            message:
              "The bucket you specified could not be accessed. Please check that the bucket is specified in your SOURCE_BUCKETS."
          };
        }
      } else {
        // Try to use the default image source bucket env var
        const sourceBuckets = this.getAllowedSourceBuckets();
        return sourceBuckets[0];
      }
    } else if (requestType === "Thumbor" || requestType === "Custom") {
      // Use the default image source bucket env var
      const sourceBuckets = this.getAllowedSourceBuckets();
      return sourceBuckets[0];
    } else {
      throw {
        status: 404,
        code: "ImageBucket::CannotFindBucket",
        message:
          "The bucket you specified could not be found. Please check the spelling of the bucket name in your request."
      };
    }
  }

  /**
   * Parses the edits to be made to the original image.
   * @param {string} event - Lambda request body.
   * @param {string} requestType - Image handler request type.
   */
  parseImageEdits(event, requestType) {
    if (requestType === "Default") {
      const decoded = this.decodeRequest(event);
      return decoded.edits;
    } else if (requestType === "Thumbor") {
      const thumborMapping = new ThumborMapping();
      thumborMapping.process(event);
      return thumborMapping.edits;
    } else if (requestType === "Custom") {
      const thumborMapping = new ThumborMapping();
      const parsedPath = thumborMapping.parseCustomPath(event.path);
      thumborMapping.process(parsedPath);
      return thumborMapping.edits;
    } else {
      throw {
        status: 400,
        code: "ImageEdits::CannotParseEdits",
        message:
          "The edits you provided could not be parsed. Please check the syntax of your request and refer to the documentation for additional guidance."
      };
    }
  }

  parseCropping(event, requestType) {
    if (requestType === "Default") {
      const decoded = this.decodeRequest(event);
      return decoded.cropping;
    } else if (requestType === "Thumbor") {
      const thumborMapping = new ThumborMapping();
      thumborMapping.process(event);
      return thumborMapping.cropping;
    } else if (requestType === "Custom") {
      const thumborMapping = new ThumborMapping();
      const parsedPath = thumborMapping.parseCustomPath(event.path);
      thumborMapping.process(parsedPath);
      return thumborMapping.cropping;
    } else {
      throw {
        status: 400,
        code: "Cropping::CannotParseCropping",
        message:
          "The cropping you provided could not be parsed. Please check the syntax of your request and refer to the documentation for additional guidance."
      };
    }
  }

  /**
   * Parses the name of the appropriate Amazon S3 key corresponding to the
   * original image.
   * @param {String} event - Lambda request body.
   * @param {String} requestType - Type, either "Default", "Thumbor", or "Custom".
   */
  parseImageKey(event, requestType) {
    if (requestType === "Default") {
      // Decode the image request and return the image key
      const decoded = this.decodeRequest(event);
      return decoded.key;
    }

    if (requestType === "Thumbor" || requestType === "Custom") {
      let {path} = event;

      if (requestType === "Custom") {
        const matchPattern = process.env.REWRITE_MATCH_PATTERN;
        const substitution = process.env.REWRITE_SUBSTITUTION;

        if (typeof matchPattern === "string") {
          const patternStrings = matchPattern.split("/");
          const flags = patternStrings.pop();
          const parsedPatternString = matchPattern.slice(1, matchPattern.length - 1 - flags.length);
          const regExp = new RegExp(parsedPatternString, flags);
          path = path.replace(regExp, substitution);
        } else {
          path = path.replace(matchPattern, substitution);
        }
      }
      path = path
        .replace(/\/\d+x\d+:\d+x\d+\//g, "/")
        .replace(/\/(\d+|__WIDTH__)x\d+\//g, "/")
        .replace(/\/(filters|roundCrop):[^\/]+/g, "/")
        .replace(/\/fit-in\//g, "/")
        .replace(/^\/+/, "")
        .replace(/\/+/g, "/");

      if (path.match(/^\d{4}\/\d{2}\/.*\/[\w-]+\.\w+$/)) {
        path = path.replace(/(.*)\/[\w-]+(\.\w+)$/, "$1/image$2");
      }
      if (path.endsWith("/")) {
        path = path + "image.jpg"
      }
      return decodeURIComponent(path);
    }

    // Return an error for all other conditions
    throw {
      status: 404,
      code: "ImageEdits::CannotFindImage",
      message:
        "The image you specified could not be found. Please check your request syntax as well as the bucket you specified to ensure it exists."
    };
  }

  /**
   * Determines how to handle the request being made based on the URL path
   * prefix to the image request. Categorizes a request as either "image"
   * (uses the Sharp library), "thumbor" (uses Thumbor mapping), or "custom"
   * (uses the rewrite function).
   * @param {object} event - Lambda request body.
   */
  parseRequestType(event) {
    const path = event["path"];
    const matchDefault = new RegExp(/^(\/?)([0-9a-zA-Z+\/]{4})*(([0-9a-zA-Z+\/]{2}==)|([0-9a-zA-Z+\/]{3}=))?$/);
    const matchThumbor = new RegExp(/^(\/?)((fit-in)?|(filters:.+\(.?\))?|(unsafe)?).*(\.+jpg|\.+png|\.+webp|\.tiff|\.jpeg|\.svg|\.gif)$/i);
    const matchCustom = new RegExp(/(\/?)(.*)(jpg|png|webp|tiff|jpeg|svg|gif)/i);

    const definedEnvironmentVariables =
      process.env.REWRITE_MATCH_PATTERN !== "" &&
      process.env.REWRITE_SUBSTITUTION !== "" &&
      process.env.REWRITE_MATCH_PATTERN !== undefined &&
      process.env.REWRITE_SUBSTITUTION !== undefined;

    //Check if path is base 64 encoded
    let isBase64Encoded = true;
    try {
      this.decodeRequest(event);
    } catch (error) {
      isBase64Encoded = false;
    }

    if (matchDefault.test(path) && isBase64Encoded) {  // use sharp
      // use sharp
      return "Default";
    } else if (matchCustom.test(path) && definedEnvironmentVariables) {
      // use rewrite function then thumbor mappings
      return "Custom";
    } else if (matchThumbor.test(path) || path.endsWith("/")) {
      // use thumbor mappings
      return "Thumbor";
    } else {
      throw {
        status: 400,
        code: "RequestTypeError",
        message:
          "The type of request you are making could not be processed. Please ensure that your original image is of a supported file type (jpg, png, tiff, webp, svg, gif) and that your image request is provided in the correct syntax. Refer to the documentation for additional guidance on forming image requests."
      };
    }
  }

  /**
   * Parses the headers to be sent with the response.
   * @param {object} event - Lambda request body.
   * @param {string} requestType - Image handler request type.
   * @return {object} Custom headers
   */
  parseImageHeaders(event, requestType) {
    if (requestType === "Default") {
      const decoded = this.decodeRequest(event);
      if (decoded.headers) {
        return decoded.headers;
      }
    }

    return undefined;
  }

  /**
   * Decodes the base64-encoded image request path associated with default
   * image requests. Provides error handling for invalid or undefined path values.
   * @param {object} event - The proxied request object.
   */
  decodeRequest(event) {
    const path = event["path"];
    if (path !== undefined) {
      const encoded = path.charAt(0) === "/" ? path.slice(1) : path;
      const toBuffer = Buffer.from(encoded, "base64");
      try {
        // To support European characters, 'ascii' was removed.
        return JSON.parse(toBuffer.toString());
      } catch (e) {
        throw {
          status: 400,
          code: "DecodeRequest::CannotDecodeRequest",
          message:
            "The image request you provided could not be decoded. Please check that your request is base64 encoded properly and refer to the documentation for additional guidance."
        };
      }
    } else {
      throw {
        status: 400,
        code: "DecodeRequest::CannotReadPath",
        message:
          "The URL path you provided could not be read. Please ensure that it is properly formed according to the solution documentation."
      };
    }
  }

  /**
   * Returns a formatted image source bucket whitelist as specified in the
   * SOURCE_BUCKETS environment variable of the image handler Lambda
   * function. Provides error handling for missing/invalid values.
   */
  getAllowedSourceBuckets() {
    const sourceBuckets = process.env.SOURCE_BUCKETS;
    if (sourceBuckets === undefined) {
      throw {
        status: 400,
        code: "GetAllowedSourceBuckets::NoSourceBuckets",
        message:
          "The SOURCE_BUCKETS variable could not be read. Please check that it is not empty and contains at least one source bucket, or multiple buckets separated by commas. Spaces can be provided between commas and bucket names, these will be automatically parsed out when decoding."
      };
    } else {
      const formatted = sourceBuckets.replace(/\s+/g, "");
      const buckets = formatted.split(",");
      return buckets;
    }
  }

  /**
   * Return the output format depending on the accepts headers and request type
   * @param {Object} event - The request body.
   */
   getOutputFormat(event) {
    const autoWebP = process.env.AUTO_WEBP;
    const autoAvif = process.env.AUTO_AVIF;
    let accept = event.headers
      ? event.headers.Accept || event.headers.accept
      : [];
    if (autoWebP === "Yes" && accept && accept.includes("image/webp")) {
      return "webp";
    } else if (autoAvif === "Yes" && accept && accept.includes("image/avif")) {
      return "avif";
    } else if (this.requestType === "Default") {
      const decoded = this.decodeRequest(event);
      return decoded.outputFormat;
    }

    return null;
  }

  /**
   * Return the output format depending on first four hex values of an image file.
   * @param {Buffer} imageBuffer - Image buffer.
   */
   inferImageType(imageBuffer) {
    switch (imageBuffer.toString("hex").substring(0, 8).toUpperCase()) {
      case "89504E47":
        return "image/png";
      case "FFD8FFDB":
        return "image/jpeg";
      case "FFD8FFE0":
        return "image/jpeg";
      case "FFD8FFEE":
        return "image/jpeg";
      case "FFD8FFE1":
        return "image/jpeg";
      case "52494646":
        return "image/webp";
      case "49492A00":
        return "image/tiff";
      case "4D4D002A":
        return "image/tiff";
      default:
        throw {
          status: 500,
          code: "RequestTypeError",
          message:
            "The file does not have an extension and the file type could not be inferred. Please ensure that your original image is of a supported file type (jpg, png, tiff, webp, svg). Refer to the documentation for additional guidance on forming image requests.",
        };
    }
  }
}

// Exports
module.exports = ImageRequest;